fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/item/v1/bow/BowExtensions.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.item.v1.bow;
+
+import net.minecraft.entity.LivingEntity;
+import net.minecraft.entity.projectile.PersistentProjectileEntity;
+import net.minecraft.item.BowItem;
+import net.minecraft.item.ItemStack;
+
+/**
+ * An interface to implement for all custom bows in fabric. <br>
+ * Note: This is meant to be used on a BowItem class, the functionality will not work otherwise.
+ *
+ * @see SimpleBowItem
+ */
+public interface BowExtensions {
+	/**
+	 * In this method you can modify the behavior of arrows shot from your custom bow. Applies all of the vanilla arrow modifiers first.
+	 *
+	 * @param bowStack                   The ItemStack for the Bow Item
+	 * @param arrowStack                 The ItemStack for the arrows
+	 * @param user                       The user of the bow
+	 * @param pullProgress          	 The pull progress of the bow from 0.0 to 1.0
+	 * @param persistentProjectileEntity The arrow entity to be spawned
+	 */
+	void onProjectileShot(ItemStack bowStack, ItemStack arrowStack, LivingEntity user, float pullProgress, PersistentProjectileEntity persistentProjectileEntity);
```

Comment:
Still, won’t the pull progress be 1 when the projectile is shot? Isn’t the projectile supposed to shoot only when the pull progress is 1?

Reply:
no, you know how you can partially charge a bow? thats what this value is

Reply:
ah I see